cdc/sinkv2/eventsink/txn/mysql/mysql.go

```diff
@@ -273,72 +273,121 @@ func convertBinaryToString(row *model.RowChangedEvent) {
 	}
 }
 
-// TODO: Find a way to make batch delete dmls more efficient.
-func groupRowsByType(
+func (s *mysqlBackend) groupRowsByType(
 	event *eventsink.TxnCallbackableEvent,
 	tableInfo *timodel.TableInfo,
 	spiltUpdate bool,
-) (insertRows, updateRows, deleteRows []*sqlmodel.RowChange) {
+) (insertRows, updateRows, deleteRows [][]*sqlmodel.RowChange) {
+	preAllocateSize := len(event.Event.Rows)
+	if preAllocateSize > s.cfg.MaxTxnRow {
+		preAllocateSize = s.cfg.MaxTxnRow
+	}
+
+	insertRow := make([]*sqlmodel.RowChange, 0, preAllocateSize)
+	updateRow := make([]*sqlmodel.RowChange, 0, preAllocateSize)
+	deleteRow := make([]*sqlmodel.RowChange, 0, preAllocateSize)
+
 	for _, row := range event.Event.Rows {
 		convertBinaryToString(row)
+
 		if row.IsInsert() {
-			insertRows = append(
-				insertRows,
+			insertRow = append(
+				insertRow,
 				convert2RowChanges(row, tableInfo, sqlmodel.RowChangeInsert))
-		} else if row.IsDelete() {
-			deleteRows = append(
-				deleteRows,
+			if len(insertRow) >= s.cfg.MaxTxnRow {
+				insertRows = append(insertRows, insertRow)
+				insertRow = make([]*sqlmodel.RowChange, 0, preAllocateSize)
+			}
+		}
+
+		if row.IsDelete() {
+			deleteRow = append(
+				deleteRow,
 				convert2RowChanges(row, tableInfo, sqlmodel.RowChangeDelete))
-		} else if row.IsUpdate() {
+			if len(deleteRow) >= s.cfg.MaxTxnRow {
+				deleteRows = append(deleteRows, deleteRow)
+				deleteRow = make([]*sqlmodel.RowChange, 0, preAllocateSize)
+			}
+		}
+
+		if row.IsUpdate() {
 			if spiltUpdate {
-				deleteRows = append(
-					deleteRows,
+				deleteRow = append(
+					deleteRow,
 					convert2RowChanges(row, tableInfo, sqlmodel.RowChangeDelete))
-				insertRows = append(
-					insertRows,
+				if len(deleteRow) >= s.cfg.MaxTxnRow {
+					deleteRows = append(deleteRows, deleteRow)
+					deleteRow = make([]*sqlmodel.RowChange, 0, s.cfg.MaxTxnRow)
+				}
+				insertRow = append(
+					insertRow,
 					convert2RowChanges(row, tableInfo, sqlmodel.RowChangeInsert))
+				if len(insertRow) >= s.cfg.MaxTxnRow {
+					insertRows = append(insertRows, insertRow)
+					insertRow = make([]*sqlmodel.RowChange, 0, s.cfg.MaxTxnRow)
+				}
 			} else {
-				updateRows = append(
-					updateRows,
+				updateRow = append(
+					updateRow,
 					convert2RowChanges(row, tableInfo, sqlmodel.RowChangeUpdate))
+				if len(updateRow) >= s.cfg.MaxTxnRow {
+					updateRows = append(updateRows, updateRow)
+					updateRow = make([]*sqlmodel.RowChange, 0, s.cfg.MaxTxnRow)
```

Comment:
Why do we use `s.cfg.MaxTxnRow` here but use `preAllocateSize` above?

Reply:
In the most time `len(event) ` is far more less than `s.cfg.MaxTxnRow`,  it will be a waste if just use `s.cfg.MaxTxnRow`.

Reply:
Thank you for your reminder.